Direct messages

deivid-rodriguez opened the pull request:
#### :tophat: What? Why?

Implements direct messages between users.

#### :pushpin: Related Issues
- Fixes #2078.

#### :clipboard: Subtasks
- [x] Empty conversation index page.
- [x] Full conversation index page.
- [x] Empty conversation page.
- [x] Full conversation page.
- [x] Reply box.
- [x] Unread counts.
- [x] Links to the feature.
  * [x] Global menu.
  * [x] Proposal author

### :camera: Screenshots (optional)

#### Conversation list
![anotherlist](https://user-images.githubusercontent.com/2887858/32968542-61c460ea-cbc0-11e7-9bea-8319ecd1b82a.png)

#### Conversation page
![conversation](https://user-images.githubusercontent.com/2887858/32968543-61ecc044-cbc0-11e7-869a-b40d28eb19f6.png)

#### Unread counts
![unread](https://user-images.githubusercontent.com/2887858/32968544-62113e1a-cbc0-11e7-9553-9300cd2b55f8.png)

#### Link to feature
![link to feature](https://user-images.githubusercontent.com/2887858/32968877-9d0bfa4a-cbc1-11e7-89d1-24ef2b5dc0ed.png)

#### :ghost: GIF
![sumo](https://user-images.githubusercontent.com/2887858/32682101-858bc500-c651-11e7-8b21-9a4695da018e.gif)

deivid-rodriguez commented:
Hei, I'm basically done here. I posted a few screenshots!

I'm thinking of linking to the feature from proposal authors and from comment authors for now, what do you think? Also, I'm now sure why but I chose the name "chat" internally, "discussion" sound more participation-ish, right? Is it worth that I rename it?

deivid-rodriguez commented:
Adding the link to the comment component is a bit tricky, I'm tempted to call this done and ship this initial version with only a message link next to the proposal author. What do you think?

deivid-rodriguez commented:
Decided to call this "done" for a first MVP. Ready for review.

josepjaume commented on decidim-core/app/helpers/decidim/messaging/chat_helper.rb:
```diff
@@ -0,0 +1,21 @@
+module Decidim
+  module Messaging
+    module ChatHelper
+      def link_to_current_or_new_chat_with(user)
```
Can you add documentation?

josepjaume commented on decidim-core/app/models/decidim/messaging.rb:
```diff
@@ -0,0 +1,9 @@
+module Decidim
+  module Messaging
+    def self.table_name_prefix
+      "decidim_messaging_"
```
Is this needed?

deivid-rodriguez replied:
No. It simplifies associations inside the namespace so one does not need to specify class names and foreign keys if they follow conventions. But it's not used anywhere else, so I should probably remove it, right?

deivid-rodriguez replied:
Hei, so I had at look into this and it is actually necessary, so that the table names of the different models are properly resolved.

At this point we could do several things:

* Don't namespace table names. That means changing the migrations so that the tables are named with just the prefix `decidim_` instead of `decidim_messaging_`.

* Move everything to a separate engine. In that case, doing `isolate_namespace Decidim::Messaging` in the engine definition would accomplish the same thing I'm doing here.

* Leave it as it is.

I vote for the last one :)

josepjaume commented on decidim-core/app/models/decidim/messaging/chat.rb:
```diff
@@ -0,0 +1,65 @@
+
+      delegate :mark_as_read, to: :receipts
+
+      def self.start!(originator:, interlocutors:, body:)
```
Can you add documentation?

josepjaume commented on decidim-core/app/models/decidim/messaging/chat.rb:
```diff
@@ -0,0 +1,65 @@
+        chat
+      end
+
+      def self.start(originator:, interlocutors:, body:)
```
Can you add documentation?

josepjaume commented on decidim-core/app/models/decidim/messaging/chat.rb:
```diff
@@ -0,0 +1,65 @@
+        chat
+      end
+
+      def add_message(sender:, body:)
```
Can you add documentation?

josepjaume commented on decidim-core/app/models/decidim/messaging/chat.rb:
```diff
@@ -0,0 +1,65 @@
+        message
+      end
+
+      def interlocutors(user)
```
Can you add documentation?

josepjaume commented on decidim-core/app/models/decidim/messaging/chat.rb:
```diff
@@ -0,0 +1,65 @@
+        participants.reject { |participant| participant.id == user.id }
+      end
+
+      def last_message
```
Can you add documentation?

josepjaume commented on decidim-core/app/models/decidim/messaging/chat.rb:
```diff
@@ -0,0 +1,65 @@
+        messages.last
+      end
+
+      def unread_count(user)
```
Can you add documentation?

josepjaume commented on decidim-core/app/models/decidim/messaging/message.rb:
```diff
@@ -0,0 +1,37 @@
+
+module Decidim
+  module Messaging
+    class Message < ApplicationRecord
```
Missing docs

josepjaume commented on decidim-core/app/models/decidim/messaging/message.rb:
```diff
@@ -0,0 +1,37 @@
+
+      validate :sender_is_participant
+
+      def envelope_for(recipients)
```
Missing docs

josepjaume commented on decidim-core/app/models/decidim/messaging/participation.rb:
```diff
@@ -0,0 +1,20 @@
+
+module Decidim
+  module Messaging
+    class Participation < ApplicationRecord
```
Missing docs

josepjaume commented on decidim-core/app/models/decidim/messaging/receipt.rb:
```diff
@@ -0,0 +1,22 @@
+
+module Decidim
+  module Messaging
+    class Receipt < ApplicationRecord
```
Missing docs

deivid-rodriguez commented:
Ooops, I forgot to add docs indeed! Thanks for pointing that out @josepjaume! I'll add them in a bit.

Regarding the other stuff:

* Yes, I thought about email notifications but I was expecting to leave and discuss that in another PR. Maybe one email notification per message is too much, maybe we should make it configurable... I prefer to leave that discussion for another time and leave it as it is for now. I added "In-app" notifications and unread counts so the feature is still useful as long as you're engaged in the platform. :)

* Regarding the public profile page, yes, that sounds like the best idea. But is there a design mockup or something at least? Again, it seems to me like something different enough to warrant a separate PR.

josepjaume commented:
@deivid-rodriguez I agree on that the profile page is something to discuss later (as it has privacy implications). Not so sure about the email thing though - I think it provides a critical value to the feature - users tend to live their lifes outside the platform, and we want to drag them in.

Maybe the @decidim/product people have something to say about it?

deivid-rodriguez commented:
Fair enough, let's wait for feedback and I'll proceed as @decidim/product thinks is best! Just note that I'm in favor of email notifications for this feature, I'm saying that this first MVP could live without them.

andreslucena commented:
@deivid-rodriguez I also think that it'd need to have email notification. 

For the moment, the best thing would be to have an email by every message. On the future we can discuss about having multiple messages by email, or having it configurable through the user profile page.

deivid-rodriguez commented:
Ok, I'll add them then!

deivid-rodriguez commented:
Two things:

* What do you think about sending an email notification only on the first unread message for each conversation? I think it's pointless if a user sends three messages in a row, to send 3 separate emails.

* Also, can you give me a raw copy of the email? I guess we should not include the body of the message, but just a link to the conversation?

deivid-rodriguez commented:
I implemented email notifications as I suggested in the previous message. Regarding the copies, I just added a raw first version, feel free to tweak it as you please.

This is ready for a second review pass!

mrcasals commented:
@deivid-rodriguez what is the "Link to feature" screenshot showing? I can't see anything different 😕

deivid-rodriguez commented:
Hahaha, yes, it's a tiny difference. It's just a little envelope next to the proposal's author that links to the conversation with her.

mrcasals commented on decidim-core/app/commands/decidim/messaging/reply_to_chat.rb:
```diff
@@ -33,6 +33,22 @@ def call
+        @message ||= chat.add_message(sender: sender, body: form.body)
+      end
+
+      def notify_interlocutors
```
Why not use the `Decidim::EventManager` here? Too verbose maybe?

deivid-rodriguez replied:
I'm not sure how this `EventManager` works, I checked how this type of emails are sent in other places and tried to do it similarly...

josepjaume replied:
Hi @deivid-rodriguez! Sorry to chime in so late. You can find an example of events working here:

https://github.com/decidim/decidim/blob/master/decidim-comments/app/commands/decidim/comments/create_comment.rb#L46

and here:

https://github.com/decidim/decidim/blob/master/decidim-comments/app/events/decidim/comments/comment_created_event.rb

Basically, an `Event` is an abstraction that creates a notification or sends an email behind the scenes, depending on which module is included (`Decidim::Events::EmailEvent` or `Decidim::Events::NotificationEvent`).

Do you think you could switch to use that? It's a really good use case for notifications :)

deivid-rodriguez replied:
Ok, will do!

deivid-rodriguez replied:
Thanks for the pointers, by the way. It makes sense to use that indeed.

deivid-rodriguez replied:
Ok, so I had a look at this and I'd rather leave it as it is because:

* There's no other example where these `Event`'s  don't create a notification but only send emails. Other places where only emails need to be sent (joining meetings, devise, report/moderation email alerts) use mailers directly.
* Current `EmailNotificationGenerator` does not support conditionally sending emails upon certain conditions on the recipients and the base resource, which I'm doing here.
* I'm not sure it would make the code better in this case. It introduces an indirection that feels like unnecesarily complicating things to me.

deivid-rodriguez replied:
After a second thought, I guess I can do the recipient filtering before publishing the event, so the second point wouldn't be a blocker. But I still think this introduces an unnecessary level of complexity for little gain. Anyways, let me know and I'll update this as you please.

josepjaume replied:
Well, the main idea here is to have a visual indication somewhere that you have unread messages, as well as receiving an email notification. But now that I give it a second thought, maybe the nature of that notification is different. 

Facebook, for example, uses two different icons in the top bar to indicate unread conversations and *events*. Maybe we should treat them differently and do something similar in `Decidim`?

We can of course do this in a further PR, but I think it's quite valuable to have this on the first iteration.

To be clear, I mean something like:

![image](https://user-images.githubusercontent.com/111746/33422258-619d203a-d5b5-11e7-92b1-d5de5e269f75.png)

deivid-rodriguez replied:
@josepjaume Note that I'm already doing that in this PR. There's a separate indicator for messages in the top menu that works just like the notification indicator: it gets hightlighted if there are new unread messages. Have a look at the screenshots!

deivid-rodriguez commented on decidim-core/spec/helpers/decidim/datetime_helper_spec.rb:
```diff
@@ -0,0 +1,49 @@
+        let(:date) { Time.zone.now - 1.year }
+
+        it "returns the full date" do
+          expect(helper.simple_date(date)).to eq("02.06.44")
```
Not sure, I copied it from my telegram list :rofl:. It uses different formats according to how close the message is in time.

mrcasals replied:
I mean the case when the date is from another year, you use `month.day.year` and I'm using `day/month/year` in my PR

deivid-rodriguez replied:
About the order, I think it's a confusion between british notation (little endian) and american notation (middle endian). Yeah, we should clarify which type of English `en.yml` is meant to hold for us. I vote for the british version which is more readable for us. If you agree, I'll change it!

mrcasals replied:
I'd go with British version, yeah 😁

deivid-rodriguez replied:
I fixed it, nice catch by the way!

deivid-rodriguez commented on decidim-core/app/controllers/decidim/messaging/chats_controller.rb:
```diff
@@ -7,6 +7,7 @@ class ChatsController < Decidim::ApplicationController
       helper Decidim::DatetimeHelper
 
       before_action :authenticate_user!
+      before_action :load_chat, only: :show
```
Yeah, you're right! And it's consistent with other controllers, so let me change that!

deivid-rodriguez commented:
Fixed the issue with the `chat` helper and rebased the PR!

mrcasals commented:
@deivid-rodriguez looks good to me! Let's wait until @josepjaume can review and approve it :)

mrcasals commented:
There's conflicts with the CHANGELOG 😄

xabier commented:
Sorry for the late response @decidim/developers , it might be too late, hope not:

@deivid-rodriguez:

> What do you think about sending an email notification only on the first unread message for each conversation? I think it's pointless if a user sends three messages in a row, to send 3 separate emails.

Can we make this configurable? It is a very nice idea. And I very much like the whole thing. But here is the use case and why it is hard to anticipate the result and you will understand why it is risky to commit to a strong policy with email notifications:

* Right now the platform does not have a massive daily use anywhere because there is no recurring activity. It is costly for cities to open participatory processes, we don't have yet any self-organized space for public discussion enabled, so participants only come to the platform (if they do!) when they get a call to action from elsewhere (e.g. they receive a newsletter, they get attention captured in social networks, they get the information somewhere else, for instance, in a poster in the street).

* We have a strong privacy policy, so we don't want user email addresses to be exported or accesible to admins, therefore we need a direct message system that makes it possible for admins to contact users. Otherwise it is very frustrating, you want to contact someone because s/he did a great proposal so help him/her improve and you have no means to contact. Now we do. But still, people don't get into the platform often enough. So what we need right now is to be able to use this direct message as a substitute for email (so notifications for each message, not just beginning of thread) is important. As the use extends, notifications get richer, activity gets more intense we might what only to call attention by email occasionaly.

Regarding naming:

* I think direct-message was a good one, why should we move to "chat"? because it is shorter? just asking. I don't have a strong opinion right now but chats scare the shit out of people after the Whatsapp epidemy and direct message seems more personal and serious. Just thinking outloud, looks to me that Direct Message or Private Message is more appropriate.

xabier commented:
And congrats for this feature! it looks really great!

deivid-rodriguez commented:
@xabier I'm not sure you understood the email notifications I implemented here. It's not one notification per thread, but one notification per "group of unread messages in a thread". Say my interlocutor sends me three messages in a row, I'll get one email notification telling me "You have new messages in your decidim inbox". Then I log in and read the new messages. If after that I get more messages, I'll get one more email notification telling me that I have more new messages. Basically I'm removing "duplicate" emails that would give the exact same information to the recipient, and thus don't add any value.

deivid-rodriguez commented:
Regarding naming, I initially chose "chat" because it was more concise indeed. But then I realized I don't like it.

Direct message is not a good name either because:
* The current "messages" are "multi-recipient". We're only creating them with a single recipient, but internally the model does not impose any limitation on the number of recipients. So "direct" seems not appropriate.
* A message represent a single "text baloon" in a "chat". So it's not really an alternative name for "chat".

As alternatives for "chat", I'd propose "discussion" or "conversation".

Finally, note that as usual the name in the UI is fully configurable, this discussion is about internal naming, although of course it's good that those match as much as possible

deivid-rodriguez commented:
Rebased without changes for now. Open to do any renaming we agree on!

josepjaume commented:
I like *Discussion*, as it implies two or more people talking to reach a conclusion. This taxonomy escalates to possible future usages like *public* discussions, open to any participant. *Conversation* seems like a closed, casual interaction between two people. I had similar concerns about using *Chat*. What do you think?

mrcasals commented:
@deivid-rodriguez conflicts in the core English locale file!

xabier commented:
_Discussion_s are what needs to happen in the comments component. This feature is to directly address a participant, despite the fact you can send messages many-to-one. We don't want to call this discussion. We should enforce public discussion. Maybe simply _message_ is good. @josepjaume @deivid-rodriguez how do you feel with that?

josepjaume commented:
I think the issue here is that `Messages` are what currently happens inside a `Chat`. If `Discussions` are something the platform already address in a different way, then I'd go for `Conversation`. A conversation is of *private* nature, and can potentially happen between more than two people.

Agree @xabier ? To sum up: A user creates a conversation with another user and sends a message. Then the other user can send other messages back, or summon other uses to the conversation.

josepjaume commented:
By the way @xabier , what do you think about this as a notification UX for messages? https://github.com/decidim/decidim/pull/2186#discussion_r154016245

deivid-rodriguez commented:
After this little discussion, it seems clear that we should rename `chat` to `conversation`, right? Should I go ahead @xabier?

xabier commented:
@josepjaume @deivid-rodriguez ok then, we have a conversation module, made of messages, we use the term message for each message, and conversation for the component/module/feature

deivid-rodriguez commented:
Almost, except that `Conversation` is not the name of the module per se, but the name of the main class inside the module.

To try to further clarify this, each entry in the first screenshot represents a `Conversation`, but each entry in the second screenshot represents a `Message`.

Also, note that there's not currently a separate component/module for this feature, it lives in core under a namespace that I've called `Decidim::Messaging`.

xabier commented:
thanks for clarifying!

mrcasals commented:
So @deivid-rodriguez is this blocked by something, or can this be merged? (there are conflicts, though)

deivid-rodriguez commented:
I need to fix conflicts and do the renaming we have agreed on `chat -> conversation`. I'll do it later today.

mrcasals commented:
Awesome, thanks! We'll wait for this before releasing 0.8.0 then 😁

josepjaume commented:
Nice!
